pkg/libvmi/cloudinit.go

```diff
@@ -60,6 +62,29 @@ func WithCloudInitNoCloudNetworkData(data string) Option {
 	}
 }
 
+// WithCloudInitNoCloudEncodedNetworkData adds cloud-init no-cloud base64 encoded network data.
+func WithCloudInitNoCloudEncodedNetworkData(networkData string) Option {
+	return func(vmi *v1.VirtualMachineInstance) {
+		addDiskVolumeWithCloudInitNoCloud(vmi, cloudInitDiskName, v1.DiskBusVirtio)
+
+		volume := getVolume(vmi, cloudInitDiskName)
+		volume.CloudInitNoCloud.NetworkDataBase64 = base64.StdEncoding.EncodeToString([]byte(networkData))
+		volume.CloudInitNoCloud.NetworkData = ""
+	}
+}
+
+// WithCloudInitNoCloudNetworkDataSecret adds cloud-init no-cloud network data from secret.
+func WithCloudInitNoCloudNetworkDataSecret(secretName string) Option {
+	return func(vmi *v1.VirtualMachineInstance) {
+		addDiskVolumeWithCloudInitNoCloud(vmi, cloudInitDiskName, v1.DiskBusVirtio)
+
+		volume := getVolume(vmi, cloudInitDiskName)
+		volume.CloudInitNoCloud.NetworkDataSecretRef = &k8scorev1.LocalObjectReference{Name: secretName}
```

Comment:
Is `NetworkDataSecretRef` mutually exclusive with `NetworkData` and `NetworkDataBased64`? If so, I think you should set it to `nil` on every `WithCloudInit*`

Reply:
I think it may contradict with other cloud-init settings yes, but I need to test it so I prefer to address it in a follow up PR.

Reply:
/hold

Please reconsider. If my hunch is correct, this new function is introducing an awkward semantic.
Besides, since this changes the API of `libvmi`, I think it would be nice if its owner @EdDev takes another look.

Reply:
This PR already address more then its intent following review comments, I prefer to addressed additional stuff on a follow up PR.
If you insistent on having  these changes, I can move all changes that are not related to this PR initial intent to another PR - second and third commits - along with what you just mention.

Reply:
Nevertheless, this PR doesn't introduce the issue you mention, it just surface it.
The same issue exist also for the user-data, the API enable referring to secret (as with network-data) but non of `WithCloudInit*` overrides it.

Reply:
> Is `NetworkDataSecretRef` mutually exclusive with `NetworkData` and `NetworkDataBased64`? If so, I think you should set it to `nil` on every `WithCloudInit*`

If we follow the logic-less build pattern, I think we should leave it up to the caller to correctly build the spec. As a tester, the caller may be interested in building a spec that is indeed colliding in order to check the validation or whatever outcome this ends with.

Does this make sense?

On the other hand, this pattern of initializing the other fields has been already taken in the other options (when setting the encoded and non-encoded versions). Even here the two are initialized.
Therefore, it does make sense to perform the initialization.

It should be fixed one way or the other and I agree it cannot be deferred to later.
The simplest option would be to do the initialization, as it continues the existing pattern.
In a follow up we can consider dropping it for all.

Another problem I see is that we assume a hard-coded disk name, which made sense when serving e2e tests only, now it is a bit awkward.
If you have a good idea how to solve it, please let me know. I guess this can also be done later.

Reply:
> Nevertheless, this PR doesn't introduce the issue you mention, it just surface it. The same issue exist also for the user-data, the API enable referring to secret (as with network-data) but non of `WithCloudInit*` overrides it. I suggest addressing it and fix all cloud-init options libvmi functions on different PR.

I do not think this is correct. You have added another options to add data to cloud-init which was not there before. Now, someone may add this data and it is not in sync with the other options.
So I think it does introduce an issue, not only surface it.

This change does initialize the encoded and non-encouded versions, so I think it is obvious that the others should reset the secret field as well.

Reply:
Thanks for explaining.
I changed other `libvmi.WithCloudInit..` that set network-data to override the NetworkSecretRef field.
Also fixed the introduced function to override the correct fields.

Reply:
Thanks!

/hold cancel